graphql_compiler/query_pagination/pagination_planning.py

```diff
@@ -1,53 +1,175 @@
-        return None
+        return PaginationPlan(tuple()), [PaginationFieldNotSpecified(pagination_node)]
 
-    if is_uuid4_type(schema_info, pagination_node.name.value, pagination_field):
-        pass
-    elif is_int_field_type(schema_info, pagination_node.name.value, pagination_field):
-        quantiles = schema_info.pagination_keys.statistics.get_field_quantiles(
-            pagination_node, pagination_field
+    if is_uuid4_type(schema_info, pagination_node, pagination_field):
+        return (
+            PaginationPlan(
+                (VertexPartitionPlan((pagination_node,), pagination_field, number_of_pages),)
+            ),
+            [],
         )
-        # We make sure there's more than enough quantiles because we don't interpolate.
-        if quantiles is None or len(quantiles) < 10 * number_of_pages:
-            return None
-    else:
-        return None
+    elif field_supports_range_reasoning(schema_info, pagination_node, pagination_field):
+        quantiles = schema_info.statistics.get_field_quantiles(pagination_node, pagination_field)
 
-    return PaginationPlan([VertexPartition((pagination_node.name.value,), number_of_pages)])
+        # If the ratio len(quantiles) // number_of_pages is N, then the largest page is
+        # expected to be (N + 1) / N times larger than the smallest page. This is because
+        # if we assume that quantiles divide the domain into equal chunks, some pages will
+        # span over N + 1 chunks, and some will span over N chunks. So if N is 5,
+        # (N + 1) / N = 1.2, which is not too bad.
+        ideal_min_num_quantiles_per_page = 5
+        ideal_quantile_resolution = ideal_min_num_quantiles_per_page * number_of_pages + 1
+
+        if quantiles is None:
+            return (
+                PaginationPlan(tuple()),
+                [
+                    InsufficientQuantiles(
+                        pagination_node, pagination_field, 0, ideal_quantile_resolution
+                    )
+                ],
+            )
+        elif len(quantiles) - 1 < number_of_pages:
+            return (
+                PaginationPlan(
+                    (VertexPartitionPlan((pagination_node,), pagination_field, len(quantiles) - 1),)
+                ),
+                [
+                    InsufficientQuantiles(
+                        pagination_node, pagination_field, len(quantiles), ideal_quantile_resolution
+                    )
+                ],
+            )
+        elif len(quantiles) < ideal_quantile_resolution:
+            return (
+                PaginationPlan(
+                    (VertexPartitionPlan((pagination_node,), pagination_field, number_of_pages),)
+                ),
+                [
+                    InsufficientQuantiles(
+                        pagination_node, pagination_field, len(quantiles), ideal_quantile_resolution
+                    )
+                ],
+            )
+        else:
+            return (
```

Comment:
Some comments on what is going on in all of these if/elif/else branches would be nice. I eventually figured it out, but I've been staring at it for the last 5 minutes and the general approach we're using is fresh in my mind. In six months, I might not be able to figure out what is going on here.